src/webview/html/editHistoryHtml.js

```diff
@@ -0,0 +1,105 @@
+import template from './template';
+import type { ThemeName, MessageSnapshot, UserOrBot, Auth } from '../../types';
+import { shortTime, humanDate } from '../../utils/date';
+import css from '../css/css';
+import script from '../js/script';
+import * as logging from '../../utils/logging';
+
+const TOPIC_EDITED = 'TOPIC_EDITED';
+const CONTENT_EDITED = 'CONTENT_EDITED';
+const BOTH_EDITED = 'BOTH_EDITED';
+const NOT_EDITED = 'NOT_EDITED';
+
+const renderSnapshot = (snapshot: MessageSnapshot, usersById: Map<number, UserOrBot>): string => {
+  const user = usersById.get(snapshot.user_id);
+  const date = new Date(snapshot.timestamp * 1000);
+  const editedTime = `${humanDate(date)} ${shortTime(date)}`;
+  let displayedContent: string = '';
+  let tag: string = '';
+  let edit_flag;
+
+  if (snapshot.prev_topic === undefined && snapshot.prev_content === undefined) {
+    edit_flag = NOT_EDITED;
+  } else if (snapshot.prev_content !== undefined && snapshot.prev_topic === undefined) {
+    edit_flag = CONTENT_EDITED;
+  } else if (snapshot.prev_content === undefined && snapshot.prev_topic !== undefined) {
+    edit_flag = TOPIC_EDITED;
+  } else {
+    edit_flag = BOTH_EDITED;
+  }
+
+  switch (edit_flag) {
+    case NOT_EDITED: {
+      tag += template`<span class="message-tag">Original message</span>`;
+      displayedContent = snapshot.rendered_content;
+      break;
+    }
+
+    case TOPIC_EDITED: {
+      tag += template`<span class="message-tag">Topic edited by ${
+        user ? user.full_name : 'unknown user'
+      }</span>`;
+
+      // prev_topic is always present for this case. $FlowFixMe.
+      displayedContent = template`Topic:
+      <span class="highlight_text_inserted">${snapshot.topic}</span>
+      <span class="highlight_text_deleted">${snapshot.prev_topic}</span>`;
+      break;
+    }
+
+    case CONTENT_EDITED: {
+      tag += template`<span class="message-tag">Edited by ${
+        user ? user.full_name : 'unknown user'
+      }</span>`;
+
+      // If prev_content is present, then so is content_html_diff. $FlowFixMe.
+      displayedContent = snapshot.content_html_diff;
+      break;
+    }
+
+    case BOTH_EDITED: {
+      tag += template`<span class="message-tag">Edited by ${
+        user ? user.full_name : 'unknown user'
+      }</span>`;
+
+      // prev_topic and content_html_diff are always present for this case. $FlowFixMe.
+      displayedContent = template`Topic:
+      <span class="highlight_text_inserted">${snapshot.topic}</span>
+      <span class="highlight_text_deleted">${snapshot.prev_topic}</span>
+      <br/>$!${snapshot.content_html_diff}`;
+      break;
+    }
+
+    default:
+      logging.error('Invalid edit_flag', { editFlag: edit_flag });
+      break;
+  }
+
+  return template`
+    <div class="edit-history-block">
+      <div class="static-timestamp">${editedTime}</div>
+      <span class="edit-history-content">$!${displayedContent}</span>
+      <div class="message-tags">$!${tag}</div>
```

Comment:
I appreciate the careful use of ``template`…` ``, `${…}`, and `$!${…}`! You may already know that this is an important surface area for preventing HTML injections.

Which we actually had, until a couple of years ago -- the day I discovered that, I [fixed it](https://github.com/zulip/zulip-mobile/commit/994145383f9478e1e2ae1b40c9b77f222ebe72d7) and then built this little templating/escaping system to help us prevent it ever happening again.

I've looked closely at these templates, and I believe everything gets escaped correctly.

A couple of things I try to do to help keep it that way, as people make changes in the future:
* Background ideas:
  * Conceptually there are two types here: text string, and HTML string. Sadly we haven't taught the type-checker the difference -- they're both `string` -- so we have to keep track of the distinction ourselves.
  * The ultimate return value is an HTML string. Most of the inputs are text strings, and in general any string that isn't explicitly meant to be used as raw HTML is a text string. The only way to get an HTML string from a text string is to hand it to `template`; and an interpolation in `template` should have `$!$` if it's an HTML string and plain `$` if it's a text string.
  * So the goal is to be able to look at a use of `$` or `$!$` -- especially a use of `$!$` -- and immediately tell, without a lot of reading through other code, whether it looks right, or should be the other way instead.
* One thing that helps: For each variable that represents a fragment that'll get used in templates, give it a name that's explicit about whether it's a text string or an HTML string. Especially if it's an HTML string, make that explicit. That way when there's an interpolation like `$!${someVar}`, and it's not a name like `$!${someHtml}`, it jumps out as suspicious.
  * This is why `messageHeaderAsHtml.js` has variables `topicHtml`, for example.
  * So here, `tagHtml` would be a good name for `tag`. Similarly `displayedContentHtml`.
* Another thing that helps: make it so that every last backtick-interpolated string in the HTML-rendering code uses `template`. That way any that don't are suspicious.
  * This commit adds one non-`template` backtick-literal: `editedTime`. The behavior is fine, because it truly is a text string, and it gets duly escaped when it's used. But it'd be good to maintain that invariant, to help make any future oversights stand out. One simple way would be to just inline that expression in the place where `editedTime` is used.

For history see 994145383f9478e1e2ae1b40c9b77f222ebe72d7, where I fixed the HTML injection (in 2018-05.) Some discussion of these hygiene practices in af5a56f41d6229faa8b60b42781951892a5a70f1, a few commits later; `template` itself was introduced in the latter's parent.